src/tactic/positivity.lean

```diff
@@ -0,0 +1,343 @@
+from `ℕ` and `ℝ≥0`).
+
+-/
+
+namespace tactic
+
+/-- Given two tactics whose result is `bool × expr`, report a `bool × expr`:
+- if at least one gives `tt`, report `tt` and one of the `tt` `expr`s
+- if neither gives `tt` but at least one gives `ff`, report `ff` and one of the `ff` `expr`s
+- if both fail, fail -/
+meta def orelse' (tac1 tac2 : tactic (bool × expr)) : tactic (bool × expr) := do
+  res ← try_core tac1,
+  match res with
+  | none := tac2
+  | some res@(ff, _) := tac2 <|> pure res
+  | some res@(tt, _) := pure res
+  end
+
+namespace positivity
+
+/-! ### Core logic of the `positivity` tactic -/
+
+private lemma lt_of_le_of_eq_of_lt {α} [preorder α] {a b b' c : α} :
+  b = b' → a ≤ b' → b < c → a < c :=
+λ h1 h2 h3, lt_of_le_of_lt (by rw h1; exact h2) h3
+
+private lemma lt_of_lt_of_eq_of_le {α} [preorder α] {a b b' c : α} :
+  b = b' → a < b' → b ≤ c → a < c :=
+λ h1 h2 h3, lt_of_lt_of_le (by rw h1; exact h2) h3
+
+private lemma le_of_eq_of_le'' {α} [preorder α] {a a' b : α} : a = a' → a ≤ b → a' ≤ b :=
+λ h1 h2, le_of_eq_of_le h1.symm h2
+
+private lemma lt_of_lt_of_eq'' {α} [preorder α] {b b' a : α} : b = b' → a < b' → a < b :=
+λ h1 h2, lt_of_lt_of_eq h2 h1.symm
+
+/-- Base case of the recursive tactic `positivity`. It proves an expression `e` is
+positive/nonnegative either by `norm_num` directly on `e`, or by finding a hypothesis of the form
+`a < e` or `a ≤ e` in which `a` can be proved positive/nonnegative by `norm_num`. -/
+meta def base (e : expr) : tactic (bool × expr) :=
+orelse' (do -- try `norm_num` to prove the goal positive directly
+  (e', p) ← norm_num.derive e <|> refl_conv e,
+  e'' ← e'.to_rat,
+  typ ← infer_type e',
+  ic ← mk_instance_cache typ,
+  if e'' > 0 then do
+    (ic, p₁) ← norm_num.prove_pos ic e',
+    p ← mk_app ``lt_of_lt_of_eq'' [p, p₁],
+    pure (tt, p)
+  else if e'' = 0 then do
+    p' ← mk_app ``ge_of_eq [p], -- check this lemma is the right one!
+    pure (ff, p')
+  else failed) $ -- loop over hypotheses
+  local_context >>= list.foldl (λ tac p₃,
+    orelse' tac $ do -- if RHS of the hypothesis is the object whose positivity is sought, try
+    -- `norm_num` to prove positivity of the LHS and then apply transitivity
+      p_typ ← infer_type p₃,
+      (lo, hi, strict) ← match p_typ with
+      | `(%%lo ≤ %%hi) := pure (lo, hi, ff)
+      | `(%%hi ≥ %%lo) := pure (lo, hi, ff)
+      | `(%%lo < %%hi) := pure (lo, hi, tt)
+      | `(%%hi > %%lo) := pure (lo, hi, tt)
+      | _ := failed
+      end,
+      is_def_eq e hi,
+      (lo', p₂) ← norm_num.derive lo <|> refl_conv lo,
+      typ ← infer_type lo',
+      ic ← mk_instance_cache typ,
+      if strict then do
+        (ic, p₁) ← norm_num.prove_nonneg ic lo',
+        p ← mk_app ``lt_of_le_of_eq_of_lt [p₂, p₁, p₃],
+        pure (tt, p)
+      else do
+        z ← mk_mapp ``has_zero.zero [some typ, none], -- there was a way to get from instance cache?
+        if lo' = z then do
```

Comment:
Don't we need an application of `is_defeq` here, so that `@has_zero (add_zero_class.to_has_zero)` is equal to `@has_zero (mul_zero_class.to_has_zero)`?

Reply:
This line disappeared after [refactoring the base case](https://github.com/leanprover-community/mathlib/pull/15618#discussion_r928160787).